Extend usefulness of tip labels

jameshadfield opened the pull request:
The motivating use case is the proliferation of datasets using an accession for `node.name` but who wish to see node names of another piece of metadata. See commit messages for more details.

Using our current monkeypox/mpxv dataset with the following differences:
* `node.name` is the accession (which is the uniq ID used throughout the pipeline). This removes the need for the pipelines to use JSON-modification scripts such as `set_final_strain_name.py`.
* "strain" is added to `node_attrs` reflecting the original strain name. (Any name can be used here.)
* `"tip_label": "strain"` added as a `display_defaults`
* "strain" added to filters and colorings


On hover boxes show both the chosen tip-label and the underlying node name (accession). (On-click modals are similar)
![image](https://github.com/nextstrain/auspice/assets/8350992/4baad320-1485-4f3d-bf6f-b6b938ebe9ca)

Filtering to duplicate strain names works intuitively
![image](https://github.com/nextstrain/auspice/assets/8350992/8a18e5c6-cba9-44fa-aec7-e91674ca5830)

jameshadfield commented:
You can test the above dataset using this PR at https://nextstrain-s-nextstrain-wpreme.herokuapp.com/staging/monkeypox/mpxv/auspice-pr-1668

victorlin commented:
There's conflicts with files I've edited since this PR was opened. Addressing them with a rebase onto latest `master`.

victorlin reviewed (changes requested):
I played around on the review app. Overall nice improvements! Just a few minor things.

jameshadfield commented on src/components/tree/infoPanels/hover.js:
```diff
@@ -434,3 +431,23 @@ const HoverInfoPanel = ({
+  }
+  /* TIP */
+  return tipLabel;
+}
```
Thanks - I've fixed this by dropping it for internal nodes. This behaviour is still there for terminal _branches_ but I think that's intuitive.

jameshadfield commented on src/components/tree/infoPanels/hover.js:
```diff
@@ -399,17 +392,21 @@ const HoverInfoPanel = ({
   const node = selectedNode.node.n;
   const idxOfInViewRootNode = getIdxOfInViewRootNode(node);
-
+  displayName(node, tipLabelKey)
```
Yeah, I can't remember, but dropped now!

jameshadfield commented on src/components/tree/infoPanels/hover.js:
```diff
@@ -434,3 +431,23 @@ const HoverInfoPanel = ({
+    if (terminal) {
+      return tipLabel ? `Branch leading to ${tipLabel}` : "Terminal branch" // hover + click
+    }
+    return `Internal branch ${tipLabelKey!==strainSymbol && tipLabel ? ` (${tipLabel})` : ''}`  // branch click only
```
👍

jameshadfield commented on src/components/tree/infoPanels/click.js:
```diff
@@ -237,27 +238,22 @@ const Trait = ({node, trait, colorings, isTerminal}) => {
           <tbody>
-            {!isTip && item(t("Number of terminal tips"), node.fullTipCount)}
+            {(isTip && tipLabelKey!==strainSymbol) && item(t("Node name"), node.name)}
+            {(!isTip && node.fullTipCount>1) && item(t("Number of terminal tips"), node.fullTipCount)}
```
👍

jameshadfield commented:
I've rebased this onto master and also simplified the process by allowing any `node_attr` to be used as a strain name rather than only colourings. See https://github.com/nextstrain/mpox/pull/275 for a PR showing the corresponding changes which make use of this PR:

mpox testing datasets available at:

* clade-IIb [current Nextstrain](https://next.nextstrain.org/staging/mpox/strain-names/mpox/clade-IIb),  new Auspice: TO DO
* clade-I [current Nextstrain](https://next.nextstrain.org/staging/mpox/strain-names/mpox/clade-I),  new Auspice: TO DO
* all-clades [current Nextstrain](https://next.nextstrain.org/staging/mpox/strain-names/mpox/all-clades),  new Auspice: TO DO

joverlee521 reviewed (approved):
Changes make sense to me. 

Used https://github.com/nextstrain/zika/pull/72 to produce testing zika dataset for the [new Auspice view](https://nextstrain-s-nextstrain-lteeid.herokuapp.com/staging/zika/trials/default-tip-label/zika).

jameshadfield commented on src/util/treeJsonProcessing.js:
```diff
@@ -32,8 +36,15 @@ const processNodes = (nodes) => {
     }
     nodeNamesSeen.add(d.name);
+
+    for (const [attrKey, attrValue] of Object.entries(d.node_attrs || {})) {
+      if (typeof attrValue === 'object' && 'value' in attrValue) {
+        nodeAttrKeys.add(attrKey)
+      }
+    }
```
Yeah, that's a good point. I wish we had made the schema so that every node-attr was the same structure.

jameshadfield replied:
I'm going to refrain from addressing this here, but I do think moving all `node_attrs` values to objects will reduce complexity in Auspice a lot, see for example all [these different getters](https://github.com/nextstrain/auspice/blob/c5857225f2911091fac20395e091031c79683b11/src/util/treeMiscHelpers.js#L12-L86). We don't have to do this on the augur side (necessitating schema changes and all that), we could probably do it within [treeJsonProcessing.js](https://github.com/nextstrain/auspice/blob/master/src/util/treeJsonProcessing.js).

jameshadfield commented on src/components/controls/choose-tip-label.js:
```diff
@@ -56,10 +62,7 @@ export function collectAvailableTipLabelOptions(colorings) {
      */
     {value: 'none', label: "none"},
     {value: strainSymbol, label: "Sample Name"},
```
One oddity I noticed here is the hardcoding of `node.name` to "Sample Name" which is a little jarring when we use accessions for `node.name`.

victorlin commented on src/components/tree/infoPanels/helpers.js:
```diff
@@ -0,0 +1,23 @@
+import { getTraitFromNode } from "../../../util/treeMiscHelpers";
+  import { numericToCalendar } from "../../../util/dateHelpers";
```
Remove extra indent:

```suggestion
import { getTraitFromNode } from "../../../util/treeMiscHelpers";
import { numericToCalendar } from "../../../util/dateHelpers";
```

jameshadfield replied:
Fixed in [force push](https://github.com/nextstrain/auspice/compare/962a2e24e46a5a6f9b28a9157deebd662d9d08db..399c50181ab1ed35a88c1a6abc95b91193502eb0)

victorlin commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+* Any `node_attr` in the tree can be used as a tip label, as well as the special-cases of strain-name and "none".
+  Previously we only allowed valid colorings.
+  ([#1668](https://github.com/nextstrain/auspice/pull/1668))
+* The specified tip label is surfaced more prominently within the the on-hover info boxes & on-click modals.
+  ([PR #1668](https://github.com/nextstrain/auspice/pull/1668))
```
Keep consistent:

```suggestion
* Any `node_attr` in the tree can be used as a tip label, as well as the special-cases of strain-name and "none".
  Previously we only allowed valid colorings.
  ([#1668](https://github.com/nextstrain/auspice/pull/1668))
* The specified tip label is surfaced more prominently within the the on-hover info boxes & on-click modals.
  ([#1668](https://github.com/nextstrain/auspice/pull/1668))
```

jameshadfield replied:
Fixed in [force push](https://github.com/nextstrain/auspice/compare/962a2e24e46a5a6f9b28a9157deebd662d9d08db..399c50181ab1ed35a88c1a6abc95b91193502eb0)

victorlin commented on src/components/tree/infoPanels/hover.js:
```diff
@@ -267,14 +262,13 @@ const BranchMutations = ({node, geneSortFn, observedMutations, t}) => {
+const BranchDescendants = ({node, t, tipLabelKey}) => {
   const [name, value] = node.fullTipCount === 1 ?
-    [t("Branch leading to"), node.name] :
+    [nodeDisplayName(node, tipLabelKey, true), ""] :
     [t("Number of descendants")+":", node.fullTipCount];
```
This is displaying `undefined`:

![image](https://github.com/user-attachments/assets/475d67b5-aece-46c1-b9d5-5b17501ad8e2)

Fix is simple:

```suggestion
    [nodeDisplayName(t, node, tipLabelKey, true), ""] :
    [t("Number of descendants")+":", node.fullTipCount];
```

I was confused how this could happen without any warnings or errors, and realized that [it's by design](https://stackoverflow.com/q/11107823) 😵 I'm guessing that's why it's common practice to pass multiple values as a single object with key/value pairs.

jameshadfield replied:
> why it's common practice to pass multiple values as a single object with key/value pairs

That's a useful approach when you want to have multiple optional values (positional & optional becomes painful unless it's the last parameter).

IMO the fix here is to leverage TypeScript rather than changing function signatures

Thanks for identifying the bug tho! Fixed [in force push](https://github.com/nextstrain/auspice/compare/962a2e24e46a5a6f9b28a9157deebd662d9d08db..399c50181ab1ed35a88c1a6abc95b91193502eb0)

victorlin reviewed (changes requested):
A couple small nits and a bug
